guess var_info for no_bs_start_match2 compilation errors

the-mikedavis opened the pull request:
Given some code like the following:

    trim_leading_dash1(<<$-, Rest/binary>>) -> trim_leading_dash1(Rest);
    trim_leading_dash1(Binary)              -> Binary.

The compiler will add an annotation for the {x,0} register

    {'%',{var_info,{x,0},[accepts_match_context]}}.

This annotation is not available in the disasm, so when we recompile
this function, the compiler gives an error.

This change detects which x-register should accept the match context
by peeking the first instruction after the entrypoint label. This
instruction can either be a {test,bs_start_match3,..} instruction or
bs_start_match4 (if the compiler can determine that the function will
_always_ be passed a binary), both of which contain the register
which will be matched.

closes #44
see also #36

the-mikedavis commented:
It looks like `are_comments_conflicting/2` might need an update, I think you can combine type information about a variable with the `accepts_match_context` annotation and still have a successful compile

dumbbell commented:
Thank you for the patch! It looks good to me.

> It looks like `are_comments_conflicting/2` might need an update, I think you can combine type information about a variable with the `accepts_match_context` annotation and still have a successful compile

Yes, `are_comments_conflicting/2` is overly conservative to catch real examples of conflicts which I didn't get so far. Do you know what `erlc -S` would emit in this case? I suppose that it combines all properties in the list of a single `{var_info, {x,0}, [...]}`, but I just want to make sure it doesn't emit multiple `var_info` for a single register.

dumbbell commented on src/khepri_fun.erl:
```diff
@@ -1132,10 +1168,9 @@ pass2_process_instruction(
   #state{mfa_in_progress = {Module, _, _},
-         label_map = LabelMap} = State)
-  when Select =:= select_val orelse Select =:= select_tuple_arity ->
+         label_map = LabelMap} = State) ->
```
This change looks unrelated to the patch. Was it expected?

the-mikedavis replied:
Whoops, good catch. Git fighting with the editor :upside_down_face:

the-mikedavis commented:
Where I ran into it, I think I tried having two `var_info`s for the same register in separate back-to-back instructions and I think that compiled. I'll try to dig up a good example...

the-mikedavis commented:
Looks like `erlc -S` groups together accepts_match_context and typing information:

```erl
{'%',{var_info,{x,0},[{type,{t_bs_context,1,0,0}},accepts_match_context]}}.
```

dumbbell reviewed (changes requested):
Sorry for delaying the merge, a coworker was working on the same part of the code around another compilation failure, but we decided to merge your work before she finishes.

Could you just squash the "restore select_tuple_arity pass2 case-statement rewriting" commit into the "refactor validation-failure handling clauses" commit please? Then I can merge.

the-mikedavis commented:
No worries, I'm not in a rush :)

I stumbled on some more var-info related validation failures and I have a big refactor I haven't been able to format as a nice PR yet. I'm curious to hear y'all's thoughts on this approach. I changed `pass2_process_instruction/2` into  `pass2_process_instructions/3` which recurses over the functions instead of being `lists:map`'ed. Then I have a few new clauses that try to preempt annotations like

```erl
pass2_process_instructions(
  [{get_tuple_element, Src, Element, _Dst} = Instruction | Rest],
  State,
  Result) ->
    VarInfo = {var_info, Src, [{type, {t_tuple, Element + 1, false, #{}}}]},
    Comment = {'%', VarInfo},
    pass2_process_instructions(Rest, State, [Instruction, Comment | Result]);
pass2_process_instructions(
  [{get_map_elements, _Fail, Src, {list, _}} = Instruction | Rest],
  State,
  Result) ->
    Instruction = replace_label(Instruction, 2, State),
    VarInfo = {var_info, Src, [{type, {t_map, any, any}}]},
    Comment = {'%', VarInfo},
    pass2_process_instructions(Rest, State, [Instruction, Comment | Result]);
pass2_process_instructions(
  [{put_map_assoc, _Fail, Src, _Dst, _Live, {list, _}} = Instruction | Rest],
  State,
  Result) ->
    VarInfo = {var_info, Src, [{type, {t_map, any, any}}]},
    Comment = {'%', VarInfo},
    pass2_process_instructions(Rest, State, [Instruction, Comment | Result]);
pass2_process_instructions(
  [{bs_start_match4, _Fail, _, Var, Var} = Instruction | Rest],
  State,
  Result) ->
    VarInfo = {var_info, Var, [accepts_match_context]},
    Comment = {'%', VarInfo},
    pass2_process_instructions(Rest, State, [Instruction, Comment | Result]);
pass2_process_instructions(
  [{test, bs_start_match3, _Fail, _, [Src], _Dst} = Instruction | Rest],
  State,
  Result) ->
    Instruction = replace_label(Instruction, 3, State),
    VarInfo = {var_info, Var, [accepts_match_context]},
    Comment = {'%', VarInfo},
    pass2_process_instructions(Rest, State, [Instruction, Comment | Result]);
%% ... Rest of pass2_process_instruction/2 clauses ...
pass2_process_instructions([], _State, Result) ->
    lists:reverse(Result).
```

This preempts most of the `beam_validator` errors but some more will come up, so I have a third pass for dealing with `beam_validator`:

```erl
validate_asm(Asm) ->
    case beam_validator:validate(Asm, weak) of
        ok ->
            Asm;
        {error, [{_GeneratedModuleName, ValidationFailures}]} ->
            fix_validation_failures(Asm, ValidationFailures)
    end.
```

And then I have some very ugly code in `fix_validation_failures/2` that `lists:foldl/3` over the failures and fixes two cases so far:

```erl
%% First: this case fails in functions that call a function that accepts
%% a binary match but doesn't have the `accepts_match_context' annotation
{_FailingFun, {{Call, Arity, {f, Label}}, _, no_bs_start_match2}
%% Fixed by scanning the Functions in Asm to the one with an entry label
%% matching Label and inserting
{'%', {var_info, {x, 0}, [accepts_match_context]}}
%% With an added TODO that the register is hard-coded and needs to be
%% determined somehow, and also that this validation failure shows up once
%% per caller, which means we could add many annotations where we only need
%% one.
```

```erl
%% Second: some binary matches also need type information, so this matches
%% the validation failure:
{{_Module, Name, Arity} = _FailingFun,
 {{bs_start_match_4, _Fail, _, Var, Var} = Instruction,
  _,
  {bad_type, {needed, Needed}, {actual, _Actual}}}
%% Which we fix by scanning to the function with that Name and Arity in Asm
%% and merging in this annotation before ^Instruction
{'%', {var_info, Var, [{type, Needed}]}}
%% And I say merging because usually there's already an annotation for that
%% variable with `accepts_match_context'.
```

 f y'all think this is a good approach, I can try to clean up my changes and make a PR. What do you think, is it unadvised to call out to `beam_validator:validate/2` directly like this?

dumbbell commented:
Another question: could we end up with multiple `var_info` added for the same register in the middle of a function body, if e.g. there is another `get_tuple_element` instruction?

Do you plan to add some kind of clean-up phase where `%` comments are moved back to the beginning of the function and possibly merged if there are multiple entries for the same register?

I don't know if the compiler is strict on where those `%` comments can appear.

the-mikedavis commented:
Makes sense about `beam_validator` vs. `beam_disasm` :+1:. I think the compiler propagates up the `{_, beam_validator, ValidationFailure}` (in erl 23: `{beam_validator, ValidationFailure}`) from that call to `beam_validator:validate/2` when `compile:forms/2` errors out. Seems wise to match on the output of `compile:forms/2` rather than `beam_validator:validate/2`.

With respect to the placement of `%`s: it seems like the compiler is very permissive, annotations don't need to follow a label and you can even have a `{'%',{var_info,[accepts_match_context]}}` followed by type info like `{'%',{var_info,[{type,{t_bs_context,1,0,0}}]}}` and as far as I can tell, the compiler is merging them.

Actually the reason I wanted to add annotations in `pass2_process_instructions/3` is that I noticed a `beam_validator` failure where it needed an annotation following a label _after_ entry-point of the function. I've got an example around here somewhere...

the-mikedavis commented:
Ok sorry for the huge code blocks here :sweat_smile:. Here's an excerpt I translated from the Elixir [Decimal](https://github.com/ericmj/decimal/blob/5825f2ae074b34791133937bcf4d76c3a071cb3a/lib/decimal.ex) library:

<details><summary>deci.erl</summary>

```erl
-module(deci).

-export([context/1, parse/1]).

do_precision(Sign, Digits, NumDigits, Exp1,
             Precision, Rounding) ->
    Precision2 = erlang:min(NumDigits, Precision),
    {Signif1, Remain1} = lists:split(Precision2,
                                         Digits),
    Signif2 = case 'increment?'(Rounding,
                                  Sign,
                                  Signif1,
                                  Remain1)
                    of
                    A when A =:= false orelse A =:= nil -> Signif1;
                    _ -> digits_increment(Signif1)
                end,
    Signals = case any_nonzero(Remain1) of
                     B when B =:= false orelse B =:= nil -> [rounded];
                     _ -> [inexact, rounded]
                 end,
    Exp2 = Exp1 + erlang:length(Remain1),
    Coef1 = digits_to_integer(Signif2),
    Dec = #{'__struct__' => 'Elixir.Decimal',
               sign => Sign, coef => Coef1, exp => Exp2},
    {Dec, Signals}.

'increment?'(_, _, _, []) -> false;
'increment?'(down, _, _, _) -> false;
'increment?'(up, _, _, _) -> true;
'increment?'(ceiling, Sign, _, Remain) ->
    case Sign == 1 of
        false -> false;
        true -> any_nonzero(Remain)
    end;
'increment?'(floor, Sign, _, Remain) ->
    case Sign == -1 of
        false -> false;
        true -> any_nonzero(Remain)
    end;
'increment?'(half_up, _, _, [Digit1 | _]) ->
    Digit1 >= 53;
'increment?'(half_even, _, [], [53 | Rest1]) ->
    any_nonzero(Rest1);
'increment?'(half_even, _, Signif1, [53 | Rest1]) ->
    case any_nonzero(Rest1) of
        false -> lists:last(Signif1) rem 2 == 1;
        true -> true;
        A -> erlang:error({badbool, 'or', A})
    end;
'increment?'(half_even, _, _, [Digit1 | _]) ->
    Digit1 > 53;
'increment?'(half_down, _, _, [Digit1 | Rest1]) ->
    case Digit1 > 53 of
        false ->
            case Digit1 == 53 of
                false -> false;
                true -> any_nonzero(Rest1)
            end;
        true -> true
    end.

any_nonzero(Digits) ->
    lists:any(fun (Digit1) -> Digit1 /= 48 end,
              Digits).

digits_to_integer([]) -> 0;
digits_to_integer(Digits1) ->
    erlang:list_to_integer(Digits1).

digits_increment(Digits1) ->
    digits_increment(lists:reverse(Digits1), []).
digits_increment([57 | Rest1], Acc1) ->
    digits_increment(Rest1, [48 | Acc1]);
digits_increment([Head1 | Rest1], Acc1) ->
    lists:reverse(Rest1, [Head1 + 1 | Acc1]);
digits_increment([], Acc1) -> [49 | Acc1].

precision(#{'__struct__' := 'Elixir.Decimal',
            coef := 'NaN'} =
              Num1,
          _Precision1, _Rounding1) ->
    {Num1, []};
precision(#{'__struct__' := 'Elixir.Decimal',
            coef := inf} =
              Num1,
          _Precision1, _Rounding1) ->
    {Num1, []};
precision(#{'__struct__' := 'Elixir.Decimal',
            sign := Sign1, coef := Coef1, exp := Exp1} =
              Num1,
          Precision1, Rounding1) ->
    Digits1 = erlang:integer_to_list(Coef1),
    _numDigits1 = erlang:length(Digits1),
    case _numDigits1 > Precision1 of
        false -> {Num1, []};
        true ->
            do_precision(Sign1,
                         Digits1,
                         _numDigits1,
                         Exp1,
                         Precision1,
                         Rounding1)
    end.

context(A) -> context(A, []).

context(Num1, Signals1) ->
    Context1 = 'Elixir.Decimal.Context':get(),
    {Result1, PrecSignals1} = precision(Num1,
                                             case Context1 of
                                                 #{precision := A} -> A;
                                                 A when erlang:is_map(A) ->
                                                     erlang:error({badkey,
                                                                   precision,
                                                                   A});
                                                 A -> A:precision()
                                             end,
                                             case Context1 of
                                                 #{rounding := B} -> B;
                                                 B when erlang:is_map(B) ->
                                                     erlang:error({badkey,
                                                                   rounding,
                                                                   B});
                                                 B -> B:rounding()
                                             end),
    begin
        C = Context1,
        D = put_uniq(Signals1, PrecSignals1),
        E = nil,
        F = Result1,
        case handle_error(D, E, F, C) of
            {ok, G} -> G;
            {error, H} ->
                erlang:error('Elixir.Decimal.Error':exception(H))
        end
    end.

put_uniq(List1, Elems1)
    when erlang:is_list(Elems1) ->
    'Elixir.Enum':reduce(Elems1,
                         List1,
                         fun (A, B) -> put_uniq(A, B) end);
put_uniq(List1, Elem1) ->
    case 'Elixir.Enum':'member?'(List1, Elem1) of
        A when A =:= false orelse A =:= nil ->
            [Elem1 | List1];
        _ -> List1
    end.

handle_error(Signals1, Reason1, Result1,
             Context1) ->
    Context2 = case Context1 of
                     A when A =:= false orelse A =:= nil ->
                         'Elixir.Decimal.Context':get();
                     B -> B
                 end,
    Signals2 = 'Elixir.List':wrap(Signals1),
    Flags1 = 'Elixir.Enum':reduce(Signals2,
                                    case Context2 of
                                        #{flags := C} -> C;
                                        C when erlang:is_map(C) ->
                                            erlang:error({badkey, flags, C});
                                        C -> C:flags()
                                    end,
                                    fun (D, E) -> put_uniq(E, D) end),
    'Elixir.Decimal.Context':set(Context2#{flags :=
                                                 Flags1}),
    ErrorSignal1 = 'Elixir.Enum':find(Signals2,
                                         fun (F) ->
                                                 'Elixir.Enum':'member?'(case
                                                                             Context2
                                                                             of
                                                                             #{traps
                                                                                   :=
                                                                                   G} ->
                                                                                 G;
                                                                             G
                                                                                 when
                                                                                   erlang:is_map(G) ->
                                                                                 erlang:error({badkey,
                                                                                               traps,
                                                                                               G});
                                                                             G ->
                                                                                 G:traps()
                                                                         end,
                                                                         F)
                                         end),
    case ErrorSignal1 of
        H when H =:= false orelse H =:= nil ->
            {ok, Result1};
        _ ->
            Error1 = [{signal, ErrorSignal1},
                        {reason, Reason1}],
            {error, Error1}
    end.

parse(<<"+", Rest1/binary>>) -> parse_unsign(Rest1);
parse(<<"-", Rest1/binary>>) ->
    case parse_unsign(Rest1) of
        {#{'__struct__' := 'Elixir.Decimal'} = Num1,
         Rest2} ->
            {Num1#{sign := -1}, Rest2};
        error -> error
    end;
parse(Binary1) when erlang:is_binary(Binary1) ->
    parse_unsign(Binary1).

parse_digits(Bin1) -> parse_digits(Bin1, []).

parse_digits(<<Digit1/integer, Rest1/binary>>,
             Acc1)
    when erlang:is_integer(Digit1) andalso
             Digit1 >= 48 andalso Digit1 =< 57 ->
    parse_digits(Rest1, [Digit1 | Acc1]);
parse_digits(Rest1, Acc1) ->
    {lists:reverse(Acc1), Rest1}.

parse_exp(<<E1/integer, Rest1/binary>>)
    when E1 =:= 101 orelse E1 =:= 69 ->
    case Rest1 of
        <<Sign1/integer, Rest2/binary>>
            when Sign1 =:= 43 orelse Sign1 =:= 45 ->
            {Digits1, Rest3} = parse_digits(Rest2),
            {[Sign1 | Digits1], Rest3};
        _ -> parse_digits(Rest1)
    end;
parse_exp(Bin1) -> {[], Bin1}.

parse_float(<<".", Rest1/binary>>) ->
    parse_digits(Rest1);
parse_float(Bin1) -> {[], Bin1}.

parse_unsign(<<First1/integer, Remainder1:7/binary,
               Rest1/binary>>)
    when First1 =:= 105 orelse First1 =:= 73 ->
    case 'Elixir.String':downcase(Remainder1) ==
             <<"nfinity">>
        of
        false -> error;
        true ->
            {#{'__struct__' => 'Elixir.Decimal', exp => 0,
               sign => 1, coef => inf},
             Rest1}
    end;
parse_unsign(<<First1/integer, Remainder1:2/binary,
               Rest1/binary>>)
    when First1 =:= 105 orelse First1 =:= 73 ->
    case 'Elixir.String':downcase(Remainder1) == <<"nf">>
        of
        false -> error;
        true ->
            {#{'__struct__' => 'Elixir.Decimal', exp => 0,
               sign => 1, coef => inf},
             Rest1}
    end;
parse_unsign(<<First1/integer, Remainder1:2/binary,
               Rest1/binary>>)
    when First1 =:= 110 orelse First1 =:= 78 ->
    case 'Elixir.String':downcase(Remainder1) == <<"an">>
        of
        false -> error;
        true ->
            {#{'__struct__' => 'Elixir.Decimal', exp => 0,
               sign => 1, coef => 'NaN'},
             Rest1}
    end;
parse_unsign(Bin1) ->
    {Int1, Rest1} = parse_digits(Bin1),
    {Float1, Rest2} = parse_float(Rest1),
    {Exp1, Rest3} = parse_exp(Rest2),
    case case Int1 == [] of
             false -> false;
             true -> Float1 == []
         end
        of
        false ->
            Int2 = case Int1 == [] of
                         false -> Int1;
                         true -> [48]
                     end,
            Exp2 = case Exp1 == [] of
                         false -> Exp1;
                         true -> [48]
                     end,
            Number1 = #{'__struct__' => 'Elixir.Decimal',
                          sign => 1,
                          coef =>
                              erlang:list_to_integer(erlang:'++'(Int2,
                                                                 Float1)),
                          exp =>
                              erlang:list_to_integer(Exp2) -
                                  erlang:length(Float1)},
            {Number1, Rest3};
        true -> error
    end.
```

</details>

And in asm that comes out to:

<details><summary>deci.S</summary>

```erl
{module, deci}.  %% version = 0

{exports, [{context,1},{module_info,0},{module_info,1},{parse,1}]}.

{attributes, []}.

{labels, 111}.


{function, do_precision, 6, 2}.
  {label,1}.
    {line,[{location,"test/fixtures/deci.erl",5}]}.
    {func_info,{atom,deci},{atom,do_precision},6}.
  {label,2}.
    {'%',{var_info,{x,1},[{type,{t_list,any,nil}}]}}.
    {'%',{var_info,{x,2},[{type,{t_integer,any}}]}}.
    {allocate,5,6}.
    {init_yregs,{list,[{y,0}]}}.
    {move,{x,5},{y,1}}.
    {move,{x,3},{y,2}}.
    {move,{x,1},{y,3}}.
    {move,{x,0},{y,4}}.
    {move,{x,4},{x,1}}.
    {move,{x,2},{x,0}}.
    {call_ext,2,{extfunc,erlang,min,2}}.
    {move,{y,3},{x,1}}.
    {init_yregs,{list,[{y,3}]}}.
    {line,[{location,"test/fixtures/deci.erl",8}]}.
    {call_ext,2,{extfunc,lists,split,2}}.
    {test,is_tuple,{f,6},[{x,0}]}.
    {test,test_arity,{f,6},[{x,0},2]}.
    {get_tuple_element,{x,0},0,{y,3}}.
    {get_tuple_element,{x,0},1,{y,0}}.
    {move,{y,0},{x,3}}.
    {move,{y,3},{x,2}}.
    {move,{y,4},{x,1}}.
    {move,{y,1},{x,0}}.
    {init_yregs,{list,[{y,1}]}}.
    {line,[{location,"test/fixtures/deci.erl",10}]}.
    {call,4,{f,8}}.
    {'%',{var_info,{x,0},[{type,{t_atom,[false,true]}}]}}.
    {test,is_eq_exact,{f,3},[{x,0},{atom,true}]}.
    {move,{y,3},{x,0}}.
    {init_yregs,{list,[{y,3}]}}.
    {line,[{location,"test/fixtures/deci.erl",16}]}.
    {call,1,{f,26}}.
    {move,{x,0},{y,3}}.
  {label,3}.
    {move,{y,0},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",18}]}.
    {call,1,{f,21}}.
    {'%',{var_info,{x,0},[{type,{t_atom,[false,true]}}]}}.
    {test,is_eq_exact,{f,4},[{x,0},{atom,true}]}.
    {move,{literal,[inexact,rounded]},{y,1}}.
    {jump,{f,5}}.
  {label,4}.
    {move,{literal,[rounded]},{y,1}}.
  {label,5}.
    {line,[{location,"test/fixtures/deci.erl",22}]}.
    {gc_bif,length,{f,0},0,[{y,0}],{x,0}}.
    {gc_bif,'+',{f,0},1,[{y,2},{x,0}],{y,2}}.
    {move,{y,3},{x,0}}.
    {move,{y,1},{y,3}}.
    {trim,2,3}.
    {line,[{location,"test/fixtures/deci.erl",23}]}.
    {call,1,{f,23}}.
    {put_map_assoc,{f,0},
                   {literal,#{'__struct__' => 'Elixir.Decimal'}},
                   {x,0},
                   1,
                   {list,[{atom,coef},
                          {x,0},
                          {atom,exp},
                          {y,0},
                          {atom,sign},
                          {y,2}]}}.
    {test_heap,3,1}.
    {put_tuple2,{x,0},{list,[{x,0},{y,1}]}}.
    {deallocate,3}.
    return.
  {label,6}.
    {line,[{location,"test/fixtures/deci.erl",8}]}.
    {badmatch,{x,0}}.


{function, 'increment?', 4, 8}.
  {label,7}.
    {line,[{location,"test/fixtures/deci.erl",28}]}.
    {func_info,{atom,deci},{atom,'increment?'},4}.
  {label,8}.
    {test,is_ne_exact,{f,19},[{x,3},nil]}.
    {select_val,{x,0},
                {f,7},
                {list,[{atom,ceiling},
                       {f,18},
                       {atom,down},
                       {f,19},
                       {atom,floor},
                       {f,17},
                       {atom,half_down},
                       {f,14},
                       {atom,half_even},
                       {f,10},
                       {atom,half_up},
                       {f,9},
                       {atom,up},
                       {f,15}]}}.
  {label,9}.
    {test,is_nonempty_list,{f,7},[{x,3}]}.
    {get_hd,{x,3},{x,0}}.
    {bif,'>=',{f,0},[{x,0},{integer,53}],{x,0}}.
    return.
  {label,10}.
    {test,is_nil,{f,11},[{x,2}]}.
    {test,is_nonempty_list,{f,7},[{x,3}]}.
    {get_list,{x,3},{x,4},{x,5}}.
    {test,is_eq_exact,{f,11},[{x,4},{integer,53}]}.
    {move,{x,5},{x,0}}.
    {call_only,1,{f,21}}.
  {label,11}.
    {test,is_nonempty_list,{f,7},[{x,3}]}.
    {get_list,{x,3},{x,0},{x,1}}.
    {test,is_eq_exact,{f,13},[{x,0},{integer,53}]}.
    {allocate,1,3}.
    {move,{x,2},{y,0}}.
    {move,{x,1},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",46}]}.
    {call,1,{f,21}}.
    {'%',{var_info,{x,0},[{type,{t_atom,[false,true]}}]}}.
    {test,is_eq_exact,{f,12},[{x,0},{atom,true}]}.
    {deallocate,1}.
    return.
  {label,12}.
    {move,{y,0},{x,0}}.
    {trim,1,0}.
    {line,[{location,"test/fixtures/deci.erl",47}]}.
    {call_ext,1,{extfunc,lists,last,1}}.
    {gc_bif,'rem',{f,0},1,[{x,0},{integer,2}],{x,0}}.
    {bif,'=:=',{f,0},[{x,0},{integer,1}],{x,0}}.
    {deallocate,0}.
    return.
  {label,13}.
    {bif,'>',{f,0},[{x,0},{integer,53}],{x,0}}.
    return.
  {label,14}.
    {test,is_nonempty_list,{f,7},[{x,3}]}.
    {get_list,{x,3},{x,0},{x,1}}.
    {test,is_lt,{f,16},[{integer,53},{x,0}]}.
  {label,15}.
    {move,{atom,true},{x,0}}.
    return.
  {label,16}.
    {test,is_eq,{f,19},[{x,0},{integer,53}]}.
    {move,{x,1},{x,0}}.
    {call_only,1,{f,21}}.
  {label,17}.
    {test,is_eq,{f,19},[{x,1},{integer,-1}]}.
    {move,{x,3},{x,0}}.
    {call_only,1,{f,21}}.
  {label,18}.
    {test,is_eq,{f,19},[{x,1},{integer,1}]}.
    {move,{x,3},{x,0}}.
    {call_only,1,{f,21}}.
  {label,19}.
    {move,{atom,false},{x,0}}.
    return.


{function, any_nonzero, 1, 21}.
  {label,20}.
    {line,[{location,"test/fixtures/deci.erl",63}]}.
    {func_info,{atom,deci},{atom,any_nonzero},1}.
  {label,21}.
    {test_heap,{alloc,[{words,0},{floats,0},{funs,1}]},1}.
    {make_fun3,{f,110},0,0,{x,1},{list,[]}}.
    {'%',{var_info,{x,1},[{fun_type,{t_atom,[false,true]}}]}}.
    {swap,{x,0},{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",64}]}.
    {call_ext_only,2,{extfunc,lists,any,2}}.


{function, digits_to_integer, 1, 23}.
  {label,22}.
    {line,[{location,"test/fixtures/deci.erl",67}]}.
    {func_info,{atom,deci},{atom,digits_to_integer},1}.
  {label,23}.
    {test,is_nil,{f,24},[{x,0}]}.
    {move,{integer,0},{x,0}}.
    return.
  {label,24}.
    {line,[{location,"test/fixtures/deci.erl",69}]}.
    {call_ext_only,1,{extfunc,erlang,list_to_integer,1}}.


{function, digits_increment, 1, 26}.
  {label,25}.
    {line,[{location,"test/fixtures/deci.erl",71}]}.
    {func_info,{atom,deci},{atom,digits_increment},1}.
  {label,26}.
    {allocate,0,1}.
    {line,[{location,"test/fixtures/deci.erl",72}]}.
    {call_ext,1,{extfunc,lists,reverse,1}}.
    {move,nil,{x,1}}.
    {call_last,2,{f,28},0}.


{function, digits_increment, 2, 28}.
  {label,27}.
    {line,[{location,"test/fixtures/deci.erl",73}]}.
    {func_info,{atom,deci},{atom,digits_increment},2}.
  {label,28}.
    {'%',{var_info,{x,0},[{type,{t_list,any,nil}}]}}.
    {'%',{var_info,{x,1},[{type,{t_list,{t_integer,{48,48}},nil}}]}}.
    {test,is_nonempty_list,{f,30},[{x,0}]}.
    {get_list,{x,0},{x,2},{x,0}}.
    {test,is_eq_exact,{f,29},[{x,2},{integer,57}]}.
    {test_heap,2,2}.
    {put_list,{integer,48},{x,1},{x,1}}.
    {call_only,2,{f,28}}.
  {label,29}.
    {line,[{location,"test/fixtures/deci.erl",76}]}.
    {gc_bif,'+',{f,0},3,[{x,2},{integer,1}],{x,2}}.
    {test_heap,2,3}.
    {put_list,{x,2},{x,1},{x,1}}.
    {call_ext_only,2,{extfunc,lists,reverse,2}}.
  {label,30}.
    {test_heap,2,2}.
    {put_list,{integer,49},{x,1},{x,0}}.
    return.


{function, precision, 3, 32}.
  {label,31}.
    {line,[{location,"test/fixtures/deci.erl",79}]}.
    {func_info,{atom,deci},{atom,precision},3}.
  {label,32}.
    {test,is_map,{f,31},[{x,0}]}.
    {get_map_elements,{f,31},
                      {x,0},
                      {list,[{atom,coef},{x,4},{atom,'__struct__'},{x,3}]}}.
    {test,is_eq_exact,{f,31},[{x,3},{atom,'Elixir.Decimal'}]}.
    {select_val,{x,4},{f,34},{list,[{atom,'NaN'},{f,33},{atom,inf},{f,33}]}}.
  {label,33}.
    {test_heap,3,1}.
    {put_tuple2,{x,0},{list,[{x,0},nil]}}.
    return.
  {label,34}.
    {get_map_elements,{f,31},
                      {x,0},
                      {list,[{atom,sign},{x,5},{atom,exp},{x,3}]}}.
    {allocate,5,6}.
    {move,{x,3},{y,0}}.
    {move,{x,5},{y,1}}.
    {move,{x,2},{y,2}}.
    {move,{x,1},{y,3}}.
    {move,{x,0},{y,4}}.
    {move,{x,4},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",93}]}.
    {call_ext,1,{extfunc,erlang,integer_to_list,1}}.
    {line,[{location,"test/fixtures/deci.erl",94}]}.
    {gc_bif,length,{f,0},1,[{x,0}],{x,1}}.
    {test,is_lt,{f,35},[{y,3},{x,1}]}.
    {move,{y,0},{x,3}}.
    {move,{x,1},{x,2}}.
    {move,{y,3},{x,4}}.
    {move,{x,0},{x,1}}.
    {move,{y,2},{x,5}}.
    {move,{y,1},{x,0}}.
    {call_last,6,{f,2},5}.
  {label,35}.
    {test_heap,3,0}.
    {put_tuple2,{x,0},{list,[{y,4},nil]}}.
    {deallocate,5}.
    return.


{function, context, 1, 37}.
  {label,36}.
    {line,[{location,"test/fixtures/deci.erl",106}]}.
    {func_info,{atom,deci},{atom,context},1}.
  {label,37}.
    {move,nil,{x,1}}.
    {call_only,2,{f,39}}.


{function, context, 2, 39}.
  {label,38}.
    {line,[{location,"test/fixtures/deci.erl",108}]}.
    {func_info,{atom,deci},{atom,context},2}.
  {label,39}.
    {'%',{var_info,{x,1},[{type,nil}]}}.
    {allocate,3,1}.
    {init_yregs,{list,[{y,0},{y,1}]}}.
    {move,{x,0},{y,2}}.
    {line,[{location,"test/fixtures/deci.erl",109}]}.
    {call_ext,0,{extfunc,'Elixir.Decimal.Context',get,0}}.
    {move,{x,0},{y,1}}.
    {test,is_map,{f,41},[{x,0}]}.
    {get_map_elements,{f,40},{x,0},{list,[{atom,precision},{x,0}]}}.
    {move,{x,0},{y,0}}.
    {jump,{f,42}}.
  {label,40}.
    {test_heap,4,0}.
    {put_tuple2,{x,0},{list,[{atom,badkey},{atom,precision},{y,1}]}}.
    {line,[{location,"test/fixtures/deci.erl",114}]}.
    {call_ext_last,1,{extfunc,erlang,error,1},3}.
  {label,41}.
    {move,{atom,precision},{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",117}]}.
    {apply,0}.
    {move,{x,0},{y,0}}.
  {label,42}.
    {test,is_map,{f,44},[{y,1}]}.
    {get_map_elements,{f,43},{y,1},{list,[{atom,rounding},{x,2}]}}.
    {jump,{f,45}}.
  {label,43}.
    {test_heap,4,0}.
    {put_tuple2,{x,0},{list,[{atom,badkey},{atom,rounding},{y,1}]}}.
    {line,[{location,"test/fixtures/deci.erl",122}]}.
    {call_ext_last,1,{extfunc,erlang,error,1},3}.
  {label,44}.
    {move,{atom,rounding},{x,1}}.
    {move,{y,1},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",125}]}.
    {apply,0}.
    {move,{x,0},{x,2}}.
  {label,45}.
    {move,{y,0},{x,1}}.
    {move,{y,2},{x,0}}.
    {init_yregs,{list,[{y,0},{y,2}]}}.
    {line,[{location,"test/fixtures/deci.erl",110}]}.
    {call,3,{f,32}}.
    {'%',
        {var_info,
            {x,0},
            [{type,
                 {t_tuple,2,true,
                     #{1 => {t_map,any,any},
                       2 => {t_list,{t_atom,[inexact,rounded]},nil}}}}]}}.
    {get_tuple_element,{x,0},0,{y,2}}.
    {get_tuple_element,{x,0},1,{x,1}}.
    {move,nil,{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",129}]}.
    {call,2,{f,48}}.
    {move,{y,1},{x,3}}.
    {move,{y,2},{x,2}}.
    {move,{atom,nil},{x,1}}.
    {init_yregs,{list,[{y,1},{y,2}]}}.
    {line,[{location,"test/fixtures/deci.erl",132}]}.
    {call,4,{f,53}}.
    {'%',
     {var_info,
      {x,0},
      [{type,
        {t_union,none,none,none,
         [{{2,{t_atom,[error]}},
           {t_tuple,2,true,
            #{1 => {t_atom,[error]},
              2 =>
               {t_cons,
                {t_union,none,none,none,
                 [{{2,{t_atom,[reason]}},
                   {t_tuple,2,true,
                    #{1 => {t_atom,[reason]},2 => {t_atom,[nil]}}}},
                  {{2,{t_atom,[signal]}},
                   {t_tuple,2,true,#{1 => {t_atom,[signal]}}}}],
                 none},
                nil}}}},
          {{2,{t_atom,[ok]}},
           {t_tuple,2,true,#{1 => {t_atom,[ok]},2 => {t_map,any,any}}}}],
         none}}]}}.
    {get_tuple_element,{x,0},0,{x,1}}.
    {get_tuple_element,{x,0},1,{x,0}}.
    {test,is_eq_exact,{f,46},[{x,1},{atom,ok}]}.
    {deallocate,3}.
    return.
  {label,46}.
    {line,[{location,"test/fixtures/deci.erl",135}]}.
    {call_ext,1,{extfunc,'Elixir.Decimal.Error',exception,1}}.
    {call_ext_last,1,{extfunc,erlang,error,1},3}.


{function, put_uniq, 2, 48}.
  {label,47}.
    {line,[{location,"test/fixtures/deci.erl",139}]}.
    {func_info,{atom,deci},{atom,put_uniq},2}.
  {label,48}.
    {test,is_list,{f,49},[{x,1}]}.
    {test_heap,{alloc,[{words,0},{floats,0},{funs,1}]},2}.
    {make_fun3,{f,108},0,0,{x,2},{list,[]}}.
    {swap,{x,0},{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",141}]}.
    {call_ext_only,3,{extfunc,'Elixir.Enum',reduce,3}}.
  {label,49}.
    {allocate,2,2}.
    {move,{x,1},{y,0}}.
    {move,{x,0},{y,1}}.
    {line,[{location,"test/fixtures/deci.erl",145}]}.
    {call_ext,2,{extfunc,'Elixir.Enum','member?',2}}.
    {select_val,{x,0},{f,51},{list,[{atom,false},{f,50},{atom,nil},{f,50}]}}.
  {label,50}.
    {test_heap,2,0}.
    {put_list,{y,0},{y,1},{x,0}}.
    {deallocate,2}.
    return.
  {label,51}.
    {move,{y,1},{x,0}}.
    {deallocate,2}.
    return.


{function, handle_error, 4, 53}.
  {label,52}.
    {line,[{location,"test/fixtures/deci.erl",151}]}.
    {func_info,{atom,deci},{atom,handle_error},4}.
  {label,53}.
    {'%',{var_info,{x,1},[{type,{t_atom,[nil]}}]}}.
    {'%',{var_info,{x,2},[{type,{t_map,any,any}}]}}.
    {allocate,3,4}.
    {move,{x,3},{y,0}}.
    {move,{x,2},{y,1}}.
    {move,{x,0},{y,2}}.
    {select_val,{x,3},{f,55},{list,[{atom,false},{f,54},{atom,nil},{f,54}]}}.
  {label,54}.
    {init_yregs,{list,[{y,0}]}}.
    {line,[{location,"test/fixtures/deci.erl",155}]}.
    {call_ext,0,{extfunc,'Elixir.Decimal.Context',get,0}}.
    {move,{x,0},{y,0}}.
  {label,55}.
    {move,{y,2},{x,0}}.
    {init_yregs,{list,[{y,2}]}}.
    {line,[{location,"test/fixtures/deci.erl",158}]}.
    {call_ext,1,{extfunc,'Elixir.List',wrap,1}}.
    {move,{x,0},{y,2}}.
    {test,is_map,{f,57},[{y,0}]}.
    {get_map_elements,{f,56},{y,0},{list,[{atom,flags},{x,0}]}}.
    {jump,{f,58}}.
  {label,56}.
    {test_heap,4,0}.
    {put_tuple2,{x,0},{list,[{atom,badkey},{atom,flags},{y,0}]}}.
    {line,[{location,"test/fixtures/deci.erl",163}]}.
    {call_ext_last,1,{extfunc,erlang,error,1},3}.
  {label,57}.
    {move,{atom,flags},{x,1}}.
    {move,{y,0},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",164}]}.
    {apply,0}.
  {label,58}.
    {test_heap,{alloc,[{words,0},{floats,0},{funs,1}]},1}.
    {make_fun3,{f,106},0,0,{x,2},{list,[]}}.
    {move,{x,0},{x,1}}.
    {move,{y,2},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",159}]}.
    {call_ext,3,{extfunc,'Elixir.Enum',reduce,3}}.
    {test,is_map,{f,61},[{y,0}]}.
    {line,[{location,"test/fixtures/deci.erl",167}]}.
    {put_map_exact,{f,0},{y,0},{x,0},1,{list,[{atom,flags},{x,0}]}}.
    {call_ext,1,{extfunc,'Elixir.Decimal.Context',set,1}}.
    {test_heap,{alloc,[{words,1},{floats,0},{funs,1}]},0}.
    {make_fun3,{f,103},0,0,{x,1},{list,[{y,0}]}}.
    {move,{y,2},{x,0}}.
    {init_yregs,{list,[{y,0},{y,2}]}}.
    {line,[{location,"test/fixtures/deci.erl",169}]}.
    {call_ext,2,{extfunc,'Elixir.Enum',find,2}}.
    {select_val,{x,0},{f,60},{list,[{atom,false},{f,59},{atom,nil},{f,59}]}}.
  {label,59}.
    {test_heap,3,0}.
    {put_tuple2,{x,0},{list,[{atom,ok},{y,1}]}}.
    {deallocate,3}.
    return.
  {label,60}.
    {test_heap,8,1}.
    {put_tuple2,{x,0},{list,[{atom,signal},{x,0}]}}.
    {put_list,{x,0},{literal,[{reason,nil}]},{x,0}}.
    {put_tuple2,{x,0},{list,[{atom,error},{x,0}]}}.
    {deallocate,3}.
    return.
  {label,61}.
    {test_heap,3,0}.
    {put_tuple2,{x,0},{list,[{atom,badmap},{y,0}]}}.
    {line,[{location,"test/fixtures/deci.erl",167}]}.
    {call_ext_last,1,{extfunc,erlang,error,1},3}.


{function, parse, 1, 63}.
  {label,62}.
    {line,[{location,"test/fixtures/deci.erl",198}]}.
    {func_info,{atom,deci},{atom,parse},1}.
  {label,63}.
    {'%',{var_info,{x,0},[accepts_match_context]}}.
    {test,bs_start_match3,{f,68},1,[{x,0}],{x,1}}.
    {bs_get_position,{x,1},{x,0},2}.
    {test,bs_get_integer2,
          {f,67},
          2,
          [{x,1},
           {integer,8},
           1,
           {field_flags,[{anno,[198,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,2}}.
    {select_val,{x,2},{f,67},{list,[{integer,43},{f,66},{integer,45},{f,64}]}}.
  {label,64}.
    {test,bs_test_unit,{f,67},[{x,1},8]}.
    {allocate,0,2}.
    {move,{x,1},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",200}]}.
    {call,1,{f,85}}.
    {'%',
        {var_info,
            {x,0},
            [{type,
                 {t_union,
                     {t_atom,[error]},
                     none,none,
                     {t_tuple,2,true,
                         #{1 =>
                               {t_map,
                                   {t_atom,['__struct__',coef,exp,sign]},
                                   any},
                           2 => {t_bitstring,1}}},
                     none}}]}}.
    {test,is_tuple,{f,65},[{x,0}]}.
    {get_tuple_element,{x,0},0,{x,1}}.
    {get_map_elements,{f,69},{x,1},{list,[{atom,'__struct__'},{x,2}]}}.
    {test,is_eq_exact,{f,69},[{x,2},{atom,'Elixir.Decimal'}]}.
    {line,[{location,"test/fixtures/deci.erl",203}]}.
    {put_map_exact,{f,0},{x,1},{x,1},2,{list,[{atom,sign},{integer,-1}]}}.
    {test_heap,3,2}.
    {get_tuple_element,{x,0},1,{x,0}}.
    {put_tuple2,{x,0},{list,[{x,1},{x,0}]}}.
    {deallocate,0}.
    return.
  {label,65}.
    {move,{atom,error},{x,0}}.
    {deallocate,0}.
    return.
  {label,66}.
    {test,bs_test_unit,{f,67},[{x,1},8]}.
    {move,{x,1},{x,0}}.
    {call_only,1,{f,85}}.
  {label,67}.
    {bs_set_position,{x,1},{x,0}}.
    {bs_get_tail,{x,1},{x,0},2}.
    {test,is_binary,{f,62},[{x,0}]}.
    {move,{x,1},{x,0}}.
    {call_only,1,{f,85}}.
  {label,68}.
    {test,is_binary,{f,62},[{x,0}]}.
    {call_only,1,{f,85}}.
  {label,69}.
    {line,[{location,"test/fixtures/deci.erl",200}]}.
    {case_end,{x,0}}.


{function, parse_digits, 1, 71}.
  {label,70}.
    {line,[{location,"test/fixtures/deci.erl",209}]}.
    {func_info,{atom,deci},{atom,parse_digits},1}.
  {label,71}.
    {'%',{var_info,{x,0},[{type,{t_bs_context,1,0,0}},accepts_match_context]}}.
    {move,nil,{x,1}}.
    {call_only,2,{f,73}}.


{function, parse_digits, 2, 73}.
  {label,72}.
    {line,[{location,"test/fixtures/deci.erl",211}]}.
    {func_info,{atom,deci},{atom,parse_digits},2}.
  {label,73}.
    {'%',{var_info,{x,0},[{type,{t_bs_context,1,0,0}},accepts_match_context]}}.
    {'%',{var_info,{x,1},[{type,{t_list,{t_integer,{0,255}},nil}}]}}.
    {bs_start_match4,{atom,resume},2,{x,0},{x,0}}.
    {bs_get_position,{x,0},{x,2},2}.
    {test,bs_get_integer2,
          {f,74},
          3,
          [{x,0},
           {integer,8},
           1,
           {field_flags,[{anno,[211,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,3}}.
    {test,bs_test_unit,{f,74},[{x,0},8]}.
    {test,is_ge,{f,74},[{x,3},{integer,48}]}.
    {test,is_ge,{f,74},[{integer,57},{x,3}]}.
    {test_heap,2,4}.
    {put_list,{x,3},{x,1},{x,1}}.
    {call_only,2,{f,73}}.
  {label,74}.
    {allocate,2,3}.
    {move,{x,2},{y,0}}.
    {move,{x,0},{y,1}}.
    {move,{x,1},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",217}]}.
    {call_ext,1,{extfunc,lists,reverse,1}}.
    {bs_set_position,{y,1},{y,0}}.
    {bs_get_tail,{y,1},{x,1},1}.
    {test_heap,3,2}.
    {put_tuple2,{x,0},{list,[{x,0},{x,1}]}}.
    {deallocate,2}.
    return.


{function, parse_exp, 1, 76}.
  {label,75}.
    {line,[{location,"test/fixtures/deci.erl",219}]}.
    {func_info,{atom,deci},{atom,parse_exp},1}.
  {label,76}.
    {'%',{var_info,{x,0},[{type,{t_bitstring,1}},accepts_match_context]}}.
    {bs_start_match4,{atom,no_fail},1,{x,0},{x,0}}.
    {bs_get_position,{x,0},{x,1},1}.
    {test,bs_get_integer2,
          {f,80},
          2,
          [{x,0},
           {integer,8},
           1,
           {field_flags,[{anno,[219,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,2}}.
    {bs_get_position,{x,0},{x,3},3}.
    {test,bs_test_unit,{f,80},[{x,0},8]}.
    {select_val,{x,2},
                {f,80},
                {list,[{integer,69},{f,77},{integer,101},{f,77}]}}.
  {label,77}.
    {test,bs_get_integer2,
          {f,79},
          4,
          [{x,0},
           {integer,8},
           1,
           {field_flags,[{anno,[222,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,1}}.
    {select_val,{x,1},{f,79},{list,[{integer,43},{f,78},{integer,45},{f,78}]}}.
  {label,78}.
    {allocate,1,2}.
    {move,{x,1},{y,0}}.
    {line,[{location,"test/fixtures/deci.erl",224}]}.
    {call,1,{f,71}}.
    {'%',{var_info,{x,0},
                   [{type,{t_tuple,2,true,
                                   #{1 => {t_list,{t_integer,{0,255}},nil},
                                     2 => {t_bitstring,1}}}}]}}.
    {test_heap,5,1}.
    {get_tuple_element,{x,0},0,{x,1}}.
    {get_tuple_element,{x,0},1,{x,0}}.
    {put_list,{y,0},{x,1},{x,1}}.
    {put_tuple2,{x,0},{list,[{x,1},{x,0}]}}.
    {deallocate,1}.
    return.
  {label,79}.
    {bs_set_position,{x,0},{x,3}}.
    {call_only,1,{f,71}}.
  {label,80}.
    {bs_set_position,{x,0},{x,1}}.
    {bs_get_tail,{x,0},{x,0},1}.
    {test_heap,3,1}.
    {put_tuple2,{x,0},{list,[nil,{x,0}]}}.
    return.


{function, parse_float, 1, 82}.
  {label,81}.
    {line,[{location,"test/fixtures/deci.erl",230}]}.
    {func_info,{atom,deci},{atom,parse_float},1}.
  {label,82}.
    {'%',{var_info,{x,0},[{type,{t_bitstring,1}},accepts_match_context]}}.
    {bs_start_match4,{atom,no_fail},1,{x,0},{x,0}}.
    {bs_get_position,{x,0},{x,1},1}.
    {test,bs_match_string,{f,83},[{x,0},8,{string,<<".">>}]}.
    {test,bs_test_unit,{f,83},[{x,0},8]}.
    {call_only,1,{f,71}}.
  {label,83}.
    {bs_set_position,{x,0},{x,1}}.
    {bs_get_tail,{x,0},{x,0},1}.
    {test_heap,3,1}.
    {put_tuple2,{x,0},{list,[nil,{x,0}]}}.
    return.


{function, parse_unsign, 1, 85}.
  {label,84}.
    {line,[{location,"test/fixtures/deci.erl",234}]}.
    {func_info,{atom,deci},{atom,parse_unsign},1}.
  {label,85}.
    {'%',{var_info,{x,0},[{type,{t_bs_matchable,1}},accepts_match_context]}}.
    {bs_start_match4,{atom,no_fail},1,{x,0},{x,0}}.
    {bs_get_position,{x,0},{x,1},1}.
    {test,bs_get_integer2,
          {f,92},
          2,
          [{x,0},
           {integer,8},
           1,
           {field_flags,[{anno,[234,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,2}}.
    {bs_get_position,{x,0},{x,3},3}.
    {test,bs_get_binary2,
          {f,87},
          4,
          [{x,0},
           {integer,7},
           8,
           {field_flags,[{anno,[234,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,4}}.
    {test,bs_get_binary2,
          {f,87},
          5,
          [{x,0},
           {atom,all},
           8,
           {field_flags,[{anno,[234,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,5}}.
    {select_val,{x,2},
                {f,87},
                {list,[{integer,73},{f,86},{integer,105},{f,86}]}}.
  {label,86}.
    {allocate,1,6}.
    {move,{x,5},{y,0}}.
    {move,{x,4},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",237}]}.
    {call_ext,1,{extfunc,'Elixir.String',downcase,1}}.
    {test,is_eq_exact,{f,91},[{x,0},{literal,<<"nfinity">>}]}.
    {jump,{f,90}}.
  {label,87}.
    {bs_set_position,{x,0},{x,3}}.
    {test,bs_get_binary2,
          {f,92},
          3,
          [{x,0},
           {integer,2},
           8,
           {field_flags,[{anno,[246,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,3}}.
    {test,bs_get_binary2,
          {f,92},
          4,
          [{x,0},
           {atom,all},
           8,
           {field_flags,[{anno,[246,{file,"test/fixtures/deci.erl"}]},
                         unsigned,big]}],
          {x,4}}.
    {select_val,{x,2},
                {f,92},
                {list,[{integer,73},
                       {f,89},
                       {integer,78},
                       {f,88},
                       {integer,105},
                       {f,89},
                       {integer,110},
                       {f,88}]}}.
  {label,88}.
    {allocate,1,5}.
    {move,{x,4},{y,0}}.
    {move,{x,3},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",260}]}.
    {call_ext,1,{extfunc,'Elixir.String',downcase,1}}.
    {test,is_eq_exact,{f,91},[{x,0},{literal,<<"an">>}]}.
    {test_heap,3,0}.
    {put_tuple2,{x,0},
                {list,[{literal,#{'__struct__' => 'Elixir.Decimal',
                                  coef => 'NaN',exp => 0,sign => 1}},
                       {y,0}]}}.
    {deallocate,1}.
    return.
  {label,89}.
    {allocate,1,5}.
    {move,{x,4},{y,0}}.
    {move,{x,3},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",249}]}.
    {call_ext,1,{extfunc,'Elixir.String',downcase,1}}.
    {test,is_eq_exact,{f,91},[{x,0},{literal,<<"nf">>}]}.
  {label,90}.
    {test_heap,3,0}.
    {put_tuple2,{x,0},
                {list,[{literal,#{'__struct__' => 'Elixir.Decimal',
                                  coef => inf,exp => 0,sign => 1}},
                       {y,0}]}}.
    {deallocate,1}.
    return.
  {label,91}.
    {move,{atom,error},{x,0}}.
    {deallocate,1}.
    return.
  {label,92}.
    {allocate,3,2}.
    {init_yregs,{list,[{y,0},{y,1},{y,2}]}}.
    {bs_set_position,{x,0},{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",269}]}.
    {call,1,{f,71}}.
    {'%',{var_info,{x,0},
                   [{type,{t_tuple,2,true,
                                   #{1 => {t_list,{t_integer,{0,255}},nil},
                                     2 => {t_bitstring,1}}}}]}}.
    {get_tuple_element,{x,0},0,{y,2}}.
    {get_tuple_element,{x,0},1,{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",270}]}.
    {call,1,{f,82}}.
    {'%',{var_info,{x,0},
                   [{type,{t_tuple,2,true,
                                   #{1 => {t_list,{t_integer,{0,255}},nil},
                                     2 => {t_bitstring,1}}}}]}}.
    {get_tuple_element,{x,0},0,{y,1}}.
    {get_tuple_element,{x,0},1,{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",271}]}.
    {call,1,{f,76}}.
    {'%',{var_info,{x,0},
                   [{type,{t_tuple,2,true,
                                   #{1 => {t_list,{t_integer,{0,255}},nil},
                                     2 => {t_bitstring,1}}}}]}}.
    {move,{x,0},{y,0}}.
    {test,is_nil,{f,94},[{y,2}]}.
    {test,is_nil,{f,93},[{y,1}]}.
    {move,{atom,error},{x,0}}.
    {deallocate,3}.
    return.
  {label,93}.
    {move,{literal,"0"},{x,0}}.
    {jump,{f,95}}.
  {label,94}.
    {move,{y,2},{x,0}}.
  {label,95}.
    {get_tuple_element,{y,0},0,{x,1}}.
    {test,is_nil,{f,96},[{x,1}]}.
    {move,{literal,"0"},{y,2}}.
    {jump,{f,97}}.
  {label,96}.
    {move,{x,1},{y,2}}.
  {label,97}.
    {move,{y,1},{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",289}]}.
    {call_ext,2,{extfunc,erlang,'++',2}}.
    {call_ext,1,{extfunc,erlang,list_to_integer,1}}.
    {swap,{y,2},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",292}]}.
    {call_ext,1,{extfunc,erlang,list_to_integer,1}}.
    {line,[{location,"test/fixtures/deci.erl",293}]}.
    {gc_bif,length,{f,0},1,[{y,1}],{x,1}}.
    {line,[{location,"test/fixtures/deci.erl",292}]}.
    {gc_bif,'-',{f,0},2,[{x,0},{x,1}],{x,0}}.
    {get_tuple_element,{y,0},1,{x,1}}.
    {put_map_assoc,{f,0},
                   {literal,#{'__struct__' => 'Elixir.Decimal',sign => 1}},
                   {x,0},
                   2,
                   {list,[{atom,coef},{y,2},{atom,exp},{x,0}]}}.
    {test_heap,3,2}.
    {put_tuple2,{x,0},{list,[{x,0},{x,1}]}}.
    {deallocate,3}.
    return.


{function, module_info, 0, 99}.
  {label,98}.
    {line,[]}.
    {func_info,{atom,deci},{atom,module_info},0}.
  {label,99}.
    {move,{atom,deci},{x,0}}.
    {call_ext_only,1,{extfunc,erlang,get_module_info,1}}.


{function, module_info, 1, 101}.
  {label,100}.
    {line,[]}.
    {func_info,{atom,deci},{atom,module_info},1}.
  {label,101}.
    {move,{x,0},{x,1}}.
    {move,{atom,deci},{x,0}}.
    {call_ext_only,2,{extfunc,erlang,get_module_info,2}}.


{function, '-handle_error/4-fun-1-', 2, 103}.
  {label,102}.
    {line,[{location,"test/fixtures/deci.erl",170}]}.
    {func_info,{atom,deci},{atom,'-handle_error/4-fun-1-'},2}.
  {label,103}.
    {'%',{var_info,{x,1},[{type,{t_map,any,any}}]}}.
    {get_map_elements,{f,104},{x,1},{list,[{atom,traps},{x,2}]}}.
    {move,{x,0},{x,1}}.
    {move,{x,2},{x,0}}.
    {line,[{location,"test/fixtures/deci.erl",171}]}.
    {call_ext_only,2,{extfunc,'Elixir.Enum','member?',2}}.
  {label,104}.
    {test_heap,4,2}.
    {put_tuple2,{x,0},{list,[{atom,badkey},{atom,traps},{x,1}]}}.
    {line,[{location,"test/fixtures/deci.erl",181}]}.
    {call_ext_only,1,{extfunc,erlang,error,1}}.


{function, '-handle_error/4-fun-0-', 2, 106}.
  {label,105}.
    {line,[{location,"test/fixtures/deci.erl",166}]}.
    {func_info,{atom,deci},{atom,'-handle_error/4-fun-0-'},2}.
  {label,106}.
    {swap,{x,0},{x,1}}.
    {call_only,2,{f,48}}.


{function, '-put_uniq/2-fun-0-', 2, 108}.
  {label,107}.
    {line,[{location,"test/fixtures/deci.erl",143}]}.
    {func_info,{atom,deci},{atom,'-put_uniq/2-fun-0-'},2}.
  {label,108}.
    {call_only,2,{f,48}}.


{function, '-any_nonzero/1-fun-0-', 1, 110}.
  {label,109}.
    {line,[{location,"test/fixtures/deci.erl",64}]}.
    {func_info,{atom,deci},{atom,'-any_nonzero/1-fun-0-'},1}.
  {label,110}.
    {bif,'/=',{f,0},[{x,0},{integer,48}],{x,0}}.
    return.
```

</details>

If you scroll down to label 92 for example, there are some annotations that are needed for the `get_tuple_element` but that's all under a label that isn't the function's entrypoint. With `pass2_process_functions/3` as I have it, that comes out to

```erl
{'%',{var_info,{x,0}, [{type,t_tuple,1,false,#{}}]}}.
{get_tuple_element,{x,0},0,{y,2}}.
{'%',{var_info,{x,0}, [{type,t_tuple,2,false,#{}}]}}.
{get_tuple_element,{x,0},1,{x,0}}.
```

But those annotations are good enough to satisfy the validator.

the-mikedavis commented:
I have a change in progress now for adding some annotations in `pass1_process_instruction/3`: [`b3cdd95`](https://github.com/rabbitmq/khepri/commit/b3cdd954bb614f3b19ad4d38bbf8fcf2b9bbad7b). It took away all the coverage on the error path of `compile:forms/2` though, so I haven't added it here yet - I'd like to figure out some test cases first.

dumbbell commented:
Nice job! I'm looking forward to see the end result :-)
